pinot-query-planner/src/main/java/org/apache/pinot/query/QueryEnvironment.java

```diff
@@ -100,26 +111,58 @@ public class QueryEnvironment {
   private final FrameworkConfig _config;
   private final CalciteCatalogReader _catalogReader;
   private final HepProgram _optProgram;
-  private final HepProgram _traitProgram;
-
-  // Pinot extensions
-  private final TableCache _tableCache;
-  private final WorkerManager _workerManager;
+  private final Config _envConfig;
 
-  public QueryEnvironment(String database, TableCache tableCache, @Nullable WorkerManager workerManager) {
-    PinotCatalog catalog = new PinotCatalog(tableCache, database);
+  public QueryEnvironment(Config config) {
+    _envConfig = config;
+    String database = config.getDatabase();
+    PinotCatalog catalog = new PinotCatalog(config.getTableCache(), database);
     CalciteSchema rootSchema = CalciteSchema.createRootSchema(false, false, database, catalog);
     _config = Frameworks.newConfigBuilder().traitDefs().operatorTable(PinotOperatorTable.instance())
         .defaultSchema(rootSchema.plus()).sqlToRelConverterConfig(PinotRuleUtils.PINOT_SQL_TO_REL_CONFIG).build();
     _catalogReader = new CalciteCatalogReader(rootSchema, List.of(database), _typeFactory, CONNECTION_CONFIG);
     _optProgram = getOptProgram();
-    _traitProgram = getTraitProgram();
-    _tableCache = tableCache;
-    _workerManager = workerManager;
   }
 
-  private PlannerContext getPlannerContext() {
-    return new PlannerContext(_config, _catalogReader, _typeFactory, _optProgram, _traitProgram);
+  public QueryEnvironment(String database, TableCache tableCache, @Nullable WorkerManager workerManager) {
+    this(configBuilder()
+        .database(database)
+        .tableCache(tableCache)
+        .workerManager(workerManager)
+        .build());
+  }
+
+  /**
+   * Returns a planner context that can be used to either parse, explain or execute a query.
+   */
+  private PlannerContext getPlannerContext(SqlNodeAndOptions sqlNodeAndOptions) {
+    boolean useImplicitColocated;
+    PinotImplicitTableHintRule.PartitionTableFinder ptf;
+    String useImplicitColocatedOptionValue = sqlNodeAndOptions.getOptions()
+        .get(CommonConstants.Broker.Request.QueryOptionKey.IMPLICIT_COLOCATE_JOIN);
+    if (Boolean.parseBoolean(useImplicitColocatedOptionValue)) {
+      useImplicitColocated = true;
+      Objects.requireNonNull(_envConfig.getWorkerManager(), "WorkerManager is required for implicit colocated join");
+      ptf = _envConfig.getWorkerManager()::getTablePartitionInfo;
+    } else {
+      useImplicitColocated = _envConfig.useImplicitColocatedByDefault();
+      WorkerManager workerManager = _envConfig.getWorkerManager();
+      if (useImplicitColocated && workerManager != null) {
+        ptf = workerManager::getTablePartitionInfo;
+      } else {
+        ptf = PinotImplicitTableHintRule.PartitionTableFinder.disabled();
+      }
+    }
+    HepProgram traitProgram = getTraitProgram(ptf, useImplicitColocated);
```

Comment:
Suggest simplifying it by using `null` `PartitionTableFinder` to represent not enable it. This way we don't need to maintain the extra boolean, and also avoid the overhead of including a rule that is disabled

Reply:
I don't like the idea of using `null`s too much as long as we don't enforce `@Nullable` notations. Therefore what I did was to introduce the `isEnabled` method in `PartitionTableFinder`. Take a look and tell me if you prefer the new code or not

Reply:
This is similar to the debate of using `null` vs `Optional`, where one is more old-fashioned and another closer to functional programming. I personally prefer `null` for consistency and slightly better but maybe negligible performance gain (less garbage and less function invocation).
Here actually we are taking higher overhead when `workerManager` is `null` because `useImplicitColocated` can be `true` while the rule is disabled. We will still match the rule. To avoid this overhead, we can only pass in the rule and check if it is disabled before adding it into the program. Then this becomes pure preference between `null` vs `Optional`

Reply:
Oh! But that has already been changed.

Anyway I will introduce extra changes to delegate on `WorkerManager` in the next push